src/game/file_io.c

```diff
@@ -812,22 +815,85 @@ static void write_int32(FILE *fp, int value)
     fwrite(&data, 1, 4, fp);
 }
 
-static int read_compressed_chunk(FILE *fp, void *buffer, int bytes_to_read)
+static int zlib_decompress(const void *input_buffer, int input_length, void *output_buffer, int output_length)
+{
+    int ret;
+    z_stream strm;
+
+    strm.zalloc = Z_NULL;
+    strm.zfree = Z_NULL;
+    strm.opaque = Z_NULL;
+    strm.avail_in = 0;
+    strm.next_in = Z_NULL;
+    ret = inflateInit(&strm);
+    if (ret != Z_OK) {
+        return 0;
+    }
+
+    strm.avail_in = input_length;
+    strm.next_in = input_buffer;
+    strm.avail_out = output_length;
+    strm.next_out = output_buffer;
+    ret = inflate(&strm, Z_NO_FLUSH);
+    inflateEnd(&strm);
+    if (ret != Z_STREAM_END || strm.avail_out != 0) {
+        return 0;
+    }
+    return 1;
+}
+
+static int read_compressed_chunk(FILE *fp, void *buffer, int bytes_to_read, int version)
 {
     if (bytes_to_read > COMPRESS_BUFFER_SIZE) {
         return 0;
     }
     int input_size = read_int32(fp);
-    if ((unsigned int) input_size == UNCOMPRESSED) {
-        if (fread(buffer, 1, bytes_to_read, fp) != bytes_to_read) {
-            return 0;
-        }
+    if ((unsigned int) input_size == UNCOMPRESSED && fread(buffer, 1, bytes_to_read, fp) != bytes_to_read) {
```

Comment:
Careful here: I think putting both statements on a single `if` can trigger the `else` code when the input is not closed and reading is successful, which is not what we want.